### ISSUE TITLE
[ New Concept Exercise] : decorators

### ISSUE THREAD

**Issue by BethanyG** [exercism/python#2356]
This issue describes how to implement the `decorators` concept exercise for the python track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [Contributing to Exercism](https://exercism.org/docs/building) | [Exercism and GitHub](https://exercism.org/docs/building/github) | [Contributor Pull Request Guide](https://exercism.org/docs/building/github/contributors-pull-request-guide)
- [What are those Weird Task Tags about?](https://exercism.org/docs/building/product/tasks)
- [Building Language Tracks: An Overview](https://exercism.org/docs/building/tracks) 
- [What are Concepts?](https://exercism.org/docs/building/tracks/concepts)
- [Concept Exercise Specifications](https://exercism.org/docs/building/tracks/concept-exercises)
- [Concept Specifications](https://exercism.org/docs/building/tracks/concepts) 
- [Exercism Formatting and Style Guide](https://exercism.org/docs/building/markdown/style-guide)
- [Exercism Markdown Specification](https://exercism.org/docs/building/markdown/markdown)
- [Reputation](https://exercism.org/docs/using/product/reputation)  

## Goal

This concept exercise is meant to teach an understanding/creation of `decorators`  in python. 

## Learning objectives

- Review/understand more details on  `higher-order functions`  in Python
  - returning `functions` from `functions`
  - passing a `function` as an argument to another `function`
  - inner or nested `functions`
- Understand that the `decorator` form and the `@` symbols are `syntatic sugar` for making/calling `higher-order functions`
- Know that `decorators` extend the behavior of an "inner", "wrapped", or passed  `function`  without explicitly modifying it.
- Create & use simple function `decorators`
- Create & use a more "complex" function `decorator`
- Use `*args` and `**kwargs` to decorate a function with different arguments
- Understand that a `decorator` is not _required_ to wrap and modify a `function`, but can simply `return` it.

  

## Out of scope

- `comprehensions`
- `class decorators`
- `classes` *as* `decorators`
- `functools` (_this will get its own exercise_)
- `functools.wraps`
- `generators`
- `lambda`, `anonymous functions`
- `map()`, `filter()`, and `reduce()` (_these will get their own exercise_)
- nested `decorators`
- stateful `decorators`

  

## Concepts

- `decorators`
- `functions`, `higher-order functions`
- `functions as arguments`
- `functions as returns`
- `nested funcitons`
- `*args` and `**kwargs`

  

## Prerequisites

These are the concepts/concept exercises the student needs to complete/understand before solving this concept exercise.  

- `basics`
- `bools`
- `comparisons`
- `dicts`
- `dict-methods`
- `functions`
- `function-arguments`
- `higher-order-functions`
- `iteration`
- `lists`
- `list-methods`
- `numbers`
- `sequences`
- `sets`
- `strings`
- `string-methods`
- `tuples`

  

## Resources to refer to

- [Python Docs: Defining Functions](https://docs.python.org/3/tutorial/controlflow.html#defining-functions)
- [Python Docs: Term - decorator](https://docs.python.org/3/glossary.html#term-decorator)
- [Functions as Objects in Python](https://medium.com/python-pandemonium/function-as-objects-in-python-d5215e6d1b0d)
- [Composing Programs: Higher-Order Functions](https://composingprograms.com/pages/16-higher-order-functions.html)
- [Real Python: Primer on Python Decorators](https://realpython.com/primer-on-python-decorators/)
- [Dan Bader: Python Decorators: A Step-by-Step Introduction](https://dbader.org/blog/python-decorators)
- [Learn by Example: Python Decorators](https://www.learnbyexample.org/python-decorators/)
- [PEP 0318 (Python.org)](https://www.python.org/dev/peps/pep-0318/)

  

* ### Hints

  For more information on writing hints see [hints](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docshintsmd)

  * You can refer to one or more of the resources linked above, or analogous resources from a trusted source.  We prefer using links within the  [Python Docs](https://docs.python.org/3/) as the primary go-to, but other resources listed above are also good.  Please try to avoid paid or subscription-based links if possible.

* ### `links.json`

  For more information, see [concept links file](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-linksjson)

  *  The same resources listed in this issue can be used as a starting point for the [ `concepts/links.json`](https://github.com/exercism/docs/blob/main/anatomy/tracks/concepts.md#file-linksjson)  file, if it doesn't already exist.
  *  If there are particularly good/interesting information sources for this concept that extend or supplement the concept exercise material & the resources already listed -- please add them to the `links.json` document.

## Concept Description

Please see the following for more details on these files:  [concepts](https://exercism.org/docs/building/tracks/concepts) & [concept exercises](https://exercism.org/docs/building/tracks/concept-exercises)

* ### Concept `about.md`

**Concept file/issue**: [decorators directory with stubbed files](https://github.com/exercism/python/tree/main/concepts/decorators) -- **Content is TBD** and should be completed as part of this exercise creation. Decorator concept write-ups and associated files can be included in the PR for this issue, or as a separate PR linked to this issue.

  For more information, see [Concept `about.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-aboutmd)

  - This file provides information about this concept for a student who has completed the corresponding concept exercise.  It is intended as a reference for continued learning.  

* ### Concept `introduction.md`

  For more information, see [Concept `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concepts.md#file-introductionmd)

  * This can also be a summary/paraphrase of the  document listed above, and will provide a brief introduction of the concept for a student who has **not yet** completed the concept exercise.   It should contain a good summation of the concept, but not go into lots of detail.

* ### Exercise `introduction.md`

  For more information, see [**Exercise** `introduction.md`](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-docsintroductionmd)

  - This should also summarize/paraphrase  the above document, but with  enough information and examples  for the student to complete the tasks outlined in this concept exercise.

  

## Test-runner

No changes required to the [Python Test Runner](https://github.com/exercism/python-test-runner) at this time.  

## Representer

No changes required to the [Python Representer](https://github.com/exercism/python-representer/) at this time.  

## Analyzer

No changes required to the [Python Analyzer](https://github.com/exercism/python-analyzer/) at this time.  

## Exercise Metadata - Track

For more information on concept exercises and formatting for the Python track `config.json` , please see [concept exercise metadata](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#metadata).  The track `config.json` file can be found in the [root of this `Python` repo](https://github.com/exercism/python/blob/main/config.json).

You can use the below for the exercise **UUID**.  You can also generate a new  one via [exercism configlet](https://github.com/exercism/configlet), [uuidgenerator.net](https://www.uuidgenerator.net/version4), or any other favorite method.  The UUID must be a valid  [**V4 UUID**](https://en.wikipedia.org/wiki/Universally_unique_identifier).

- **Exercise UUID** : `a9e59661-c687-45fb-93e5-622ed612a060`
- **concepts** should be filled in from the Concepts section in this issue
- **prerequisites** should be filled in from the Prerequisites section in this issue  

## Exercise Metadata Files Under  `.meta/config.json`

For more information on exercise  `.meta/`  files and formatting, see [concept exercise metadata files](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#files)

- `.meta/config.json` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metaconfigjson) for the fields and formatting of this file.
- `.meta/design.md` - see [this link](https://github.com/exercism/docs/blob/main/building/tracks/concept-exercises.md#file-metadesignmd) for the formatting of this file.  Please use the **Goal**, **Learning Objectives**,**Concepts**, **Prerequisites** and , **Out of Scope**  sections  from this issue.

  

## Implementation Notes

- Code in the `.meta/examplar.py` file should **only use syntax & concepts introduced in this exercise or one of its prerequisite exercises.**  We run all our `examplar.py` files through PyLint, but do not require module docstrings.  We **do** require function docstrings similar to [PEP257](https://www.python.org/dev/peps/pep-0257/).  See [this concept exercise `exemplar.py`](https://github.com/exercism/python/blob/main/exercises/concept/meltdown-mitigation/.meta/exemplar.py) for an example.
- Please **do not use** comprehensions, generator expressions, or other syntax not previously covered.  Please also follow [PEP8](https://www.python.org/dev/peps/pep-0008/) guidelines.
- In General, tests should be written using `unittest.TestCase` and the test file should be named `<EXERCISE-NAME>_test.py`.  

    - All asserts should contain a "user friendly" failure message (_these will display on the webiste_). 
    - We use a `PyTest custom mark` to link test cases to exercise task numbers.  
    - We also use `unittest.subtest` to parameterize test input where/when needed. 
Here is an [example testfile](https://github.com/exercism/python/blob/main/exercises/concept/making-the-grade/loops_test.py) that shows all three of these in action.

- While we do use [PyTest](https://docs.pytest.org/en/stable/) as our test runner and for some implementation tests, please check with a maintainer before using  a PyTest test method, fixture,  or feature.
- Our markdown and JSON files are checked against [prettier](https://prettier.io/) .  We recommend [setting prettier up locally](https://prettier.io/docs/en/install.html) and running it prior to submitting your PR  to avoid any CI errors.

## Help

If you have any questions while implementing the exercise, please post the questions as comments in this issue, or contact one of the maintainers on our Slack channel.

**Comment by BethanyG** [comment]
We may want to consider adding [`functools.wraps()`](https://docs.python.org/3/library/functools.html#functools.wraps) to this exercise, as it is very helpful in defining decorators.  However, it does introduce the whole `functools` _module_, which may not be what we want to do here.

**Comment by doublevcodes** [comment]
Hello, I would like to contribute to exercism, can I try doing this?

**Comment by BethanyG** [comment]
@doublevcodes 👋🏽   (_Python maintainer here_) Yes!  It can be _all yours_ 😄   🌟  

I will flag it as claimed, and you can get to work.  Some of the links in here _might_ be out of date -- I will update them today.  Please don't hesitate to reach out with any questions or issues you might have.  And thank you for taking this on. 🌈

**Comment by J08K** [comment]
Heya @doublevcodes, just checking in on this issue. Haven't heard anything from you for a few days.

If you need any help or are stuck on this issue, feel free to contact me by pinging me on Github or Slack! Would also love to know how much you've done. Let me know!

**Comment by BethanyG** [comment]
@doublevcodes  -- someone else has asked if they can work on this issue, so I am un-assigning you.  Please let me know if you want to continue/start work on it, and I can re-assign you.  Thanks!  🌟

**Comment by mathstrains21** [comment]
Hi! I would love to work on this issue! I'm a little confused by what some of the text in the OP means - do I need to write concept and exercise or just exercise?

**Comment by BethanyG** [comment]
Hi @mathstrains21 !  

Thanks so much for stepping forward to work on this issue -- **we really appreciate it!** 🌟 

> _do I need to write concept and exercise or just exercise?_

Both the [concept docs](https://github.com/exercism/python/tree/main/concepts/decorators) and the concept _exercise_  itself need to be written.  However, we can deal with that in several ways:

1.  I or another maintainer can write the concept docs, if you'd like to concentrate on the exercise.
2. You can write the concept docs and leave the exercise for a maintainer or someone else to work on.
3. You can work on the exercise first, and then submit a follow-on PR for the concept docs - or vice-versa.

Just let us know how you'd like to proceed, and if you have any additional questions.  

Please also keep in mind that this exercise has prerequisites that are not yet live on the site, so it may not be published right away.  If you're looking to get something in front of users more quickly, we do have an exercise on `conditionals` that needs a re-work, as well as exercises on `string-formatting` and `complex-numbers`.  Just let us know.  😄

**Comment by mathstrains21** [comment]
Hi! I think I will do both! If I don't write good explanations etc. that is what the review process is for!

What do you think is easier to do first? I guess writing exercise first so then explanations can lead a bit from that?

Note:
I might not get much done over the next few weeks due to school! After that Christmas holidays give me loooooooooads of time :-)

**Comment by BethanyG** [comment]
> _I think I will do both! If I don't write good explanations etc. that is what the review process is for!_

_I love your "can-do" spirit!_  And yes -- as you have probably seen, the review process can be lengthy.  But it can also be very collaborative and fun -- **all of our exercises are a work in progress,** and all of them could be improved.  In fact, we're going back over all of our concept exercises over the holidays and seeing where we can make them better.

> _What do you think is easier to do first?_ 

😄     This is a personal thing.  I myself usually do the concept `about.md` first.  It gives me a chance to read up on all the details, go through a bunch of links (_which I can then use in the `links.json` file later_), and practice a bunch of things in preparation for writing the exercise.  Once I have an `about.md`, I summarize it for the concept `introduction.md`, and usually paraphrase it, shorten it, or add specific code examples to it for the **exercise** `introduction.md`. 

Usually, in the process of "researching" for the concept `about.md`, a story or problem comes to me for the exercise.  But it also depends on the concept.  Many early concepts are harder, because you cannot use code or techniques that haven't been introduced as part of the exercise prerequisites.  But `decorators` are far enough "down the tree", that you can use most common techniques in your problem setup or solution.  We also have a ["library" of exercise stories and implementations](https://github.com/exercism/docs/tree/main/building/tracks/stories) -- so there may be an exercise from `Ruby`, `Go-lang`, `JavaScript`,  `C#`, `Elixir`, or another language that could be ported or adapted for use here.

> _I might not get much done over the next few weeks due to school! After that Christmas holidays give me loooooooooads of time_ 

No worries there!  I am one of the world's slowest writers ... so even when I am working on an exercise, it takes me several weeks!  Just give us (myself or @J08K ) a ping if you get stuck, want to discuss things via chat or call, or if you decided you'd rather work on something else.  And good luck with any pending papers or tests!  🌟

**Comment by mathstrains21** [comment]
Hi, so could I confirm that this is what I should fill in:

python/config.json - add in exercise
concept/about.md - add a detailed explanation of decorators
concept/introduction.md - give some basic info about decorators
concept/links.json - provide links for further reading
concept/.meta/config.json - add blurb and myself as an author

The exercise files I have got a little confused about, so could someone summarise that for me so I understand!

**Comment by J08K** [comment]
Heya @mathstrains21,
We have documentation linked in the `## Getting started` header. This should probably tell you what you need to know. You could also watch [this video](https://www.youtube.com/watch?v=gkbBqd7hPrA) if that's more your style :). If you really need help, just message me on slack (@J08K).
Have a nice day / evening! 🐍

**Comment by mathstrains21** [comment]
Hi!

I had read that, but I was probably quite tired so didn't learn too much from it! Will review now that I have been refreshed after a good sleep!

Just to note, the what is the weird tasks tag link above points to a non-existent page. It points to /building/tasks, while the page on the docs points to /building/product/tasks!

**Comment by mathstrains21** [comment]
Hi again! Is there a recommended max line length in the source markdown files?

**Comment by BethanyG** [comment]
@mathstrains21 --

> _Just to note, the what is the weird tasks tag link above points to a non-existent page. It points to /building/tasks, while the page on the docs points to /building/product/tasks!_

Thanks for the head's up!  We keep re-organizing docs on the site, and then forgetting to update issues.  _Sigh_.

> _Is there a recommended max line length in the source markdown files?_

Nope!  But we strongly recommend the [one sentence per line](https://asciidoctor.org/docs/asciidoc-recommended-practices/#one-sentence-per-line), which can feel really confusing at first.  The docs are parsed and then re-formatted for the web site, so the EOL at the end of each sentence is stripped, but having each sentence on its own line greatly helps with tracking edits and readability of source.  

You will still want to do line breaks for paragraphs by using two spaces + a [return] followed by a blank line. Same goes for separating code blocks.  Here's an example:


<img src="https://user-images.githubusercontent.com/5923094/144717713-46ee35a9-2a24-4fe7-a515-47c2e97c49ee.png" width= 600>

### Which would render like:

This is the first sentence of the paragraph. This is the second sentence of the paragraph. Here is the final sentence of the paragraph.  

Here is the start of the second paragraph.  

```python
# Code block goes here.

```

And here is the follow-on text.

----


But @J08K and I are also happy to help review things as you go along.  You can either ping us in Slack -- or submit an in-progress PR that is flagged as `draft` with a message for feedback.  😄

**Comment by BethanyG** [comment]
For some examples of how the various pieces of a concept exercise are put together (and what a concept `about.md` looks like VS a concept `introduction.md`) check out these concept exercises:

[Guidos Gorgeous Lasagna](https://github.com/exercism/python/tree/main/exercises/concept/guidos-gorgeous-lasagna) & [basics concept](https://github.com/exercism/python/tree/main/concepts/basics)
[Inventory Management](https://github.com/exercism/python/tree/main/exercises/concept/inventory-management) & [dicts concept](https://github.com/exercism/python/tree/main/concepts/dicts)
[Cater Waiter](https://github.com/exercism/python/tree/main/exercises/concept/cater-waiter) & [sets concept](https://github.com/exercism/python/tree/main/concepts/sets)
[Little Sister's Essay](https://github.com/exercism/python/tree/main/exercises/concept/little-sisters-essay) & [string methods concept](https://github.com/exercism/python/tree/main/concepts/string-methods)

**Comment by BethanyG** [comment]
Also -- because of the prerequisites and flags for setting an exercise live -- it is probably best if @J08K or I do the edits to the track `python/config.json`.  We'll do it as a last step when we finalize your PR.  😄
